org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/typeHierarchy/TypeHierarchyView.java

```diff
@@ -11,32 +11,343 @@
+	private Label infoText;
+
+	// viewers
+	private TableViewer memberViewer;
+	protected TreeViewer treeViewer;
+
+	private volatile CompletableFuture<List<Either<SymbolInformation, DocumentSymbol>>> symbols;
+	private HashMap<URI, SymbolsContainer> cachedSymbols = new HashMap<>();
+	private IDocument document;
+	private volatile String typeName;
+
+	private final IFileBufferListener fileBufferListener = new FileBufferListenerAdapter() {
+		@Override
+		public void dirtyStateChanged(IFileBuffer buffer, boolean isDirty) {
+			if (isDirty) {
+				// check if this file has been cached:
+				var cachedSymbol = getSymbolsContainer(buffer);
+				if (cachedSymbol != null) {
+					cachedSymbol.isDirty = true;
+				}
+			}
+		}
+
+		@Override
+		public void underlyingFileMoved(IFileBuffer buffer, IPath path) {
+			var symbolsContainer = getSymbolsContainer(buffer);
+			// check if this file has been cached:
+			if (symbolsContainer != null) {
+				var file = LSPEclipseUtils.getFile(path);
+				if (file != null) {
+					//update old cache:
+					symbolsContainer.setFile(file);
+					//create new cache element under new URI
+					cachedSymbols.put(file.getLocationURI(), new SymbolsContainer(file));
+				}
+			}
+		}
+
+		private SymbolsContainer getSymbolsContainer(IFileBuffer buffer) {
+			if (buffer != null) {
+				return cachedSymbols.get(LSPEclipseUtils.toUri(buffer));
+			}
+			return null;
+		}
+
+	};
+
+	private IDocumentListener documentListener = new IDocumentListener() {
+
+		@Override
+		public void documentChanged(DocumentEvent event) {
+			var document = event.getDocument();
+			if (document != null) {
+				refreshMemberViewer(LSPEclipseUtils.getFile(document), typeName, true);
+			}
+		}
+
+		@Override
+		public void documentAboutToBeChanged(DocumentEvent event) {
+			// Do nothing
+		}
+	};
+
+	@Override
+	public void setFocus() {
+		pagebook.setFocus();
+	}
 
 	@Override
 	public void createPartControl(Composite parent) {
-		final var filteredTree = new FilteredTree(parent, SWT.BORDER, new PatternFilter(), true, false) {
+		pagebook = new PageBook(parent, SWT.NULL);
+		splitter = new SashForm(pagebook, SWT.VERTICAL);
+		splitter.setLayoutData(new GridData(GridData.FILL_BOTH));
+		splitter.addControlListener(new ControlListener() {
+			@Override
+			public void controlMoved(ControlEvent e) {
+			}
+
+			@Override
+			public void controlResized(ControlEvent e) {
+				splitter.setOrientation(getBestOrientation());
+				splitter.layout();
+			}
+		});
+
+		treeViewer = getFilteredTree(splitter).getViewer();
+		treeViewer.setContentProvider(contentProvider);
+
+		treeViewer.setLabelProvider(new TypeHierarchyItemLabelProvider());
+
+		treeViewer.setUseHashlookup(true);
+		treeViewer.getControl().setEnabled(false);
+		treeViewer.addDoubleClickListener(new IDoubleClickListener() {
+			@Override
+			public void doubleClick(final DoubleClickEvent event) {
+				var selection = ((IStructuredSelection) event.getSelection()).getFirstElement();
```

Comment:
could we have a ClassCastException here?

Reply:
I don't think so because `event.getSelection` returns a `TreeSelection` object and `TreeSelection` extends `StructuredSelection` which implements `IStructuredSelection`.